opendap_tests.py

```diff
@@ -54,18 +72,184 @@ def url_tester_ext(url_address, ext='.dmr'):
+            t = build_leaf_path(v)
+            dap_url = url_address + '.dap?dap4.ce=/' + t
+            #  print(dap_url)
+            dap_r = requests.get(dap_url)
+            if dap_r.status_code == 200:
+                # Save the response?
+                if save_all:
+                    base_name = url_address.split('/')[-1]
+                    if save:
+                        base_name = save + '/' + base_name
+                    with open(base_name + '.dap', 'w') as file:
+                        file.write(dap_r.text)
+            else:
+                print("F", end="", flush=True) if not quiet else False
+
+                tr = TestResult("fail", dap_r.status_code)
+                results[dap_url] = tr
+
+                base_name = url_address.split('/')[-1]
+                if save:
+                    base_name = save + '/' + base_name
+                with open(base_name + '.dap' + '.fail.txt', 'w') as file:
+                    file.write(f'Status: {dap_r.status_code}: {dap_r.text}')
+
+    # Ignore exception, the url_tester will return 'fail'
+    except requests.exceptions.InvalidSchema:
+        pass
+
+    return results
+
+
+def build_leaf_path (var):
+    path = var.getAttribute("name")
+    # print(path)
+    if var.parentNode.nodeName != "Dataset":
+        if var.parentNode.nodeName == "Group":
+            path = build_leaf_path(var.parentNode) + '/' + path
+        elif var.parentNode.nodeName == "Structure":
+            path = build_leaf_path(var.parentNode) + '.' + path
+        elif var.parentNode.nodeName == "Sequence":
+            path = build_leaf_path(var.parentNode) + '.' + path
+        # print(path)
+    return path
+
+
+def url_test_runner(url, dmr=True, dap=True, nc4=False):
     """
     Common access point for all the tests.
     """
-    test_results = {"dmr": url_tester_ext(url) if dmr else "NA",
-                    "dap": url_tester_ext(url, '.dap') if dap else "NA",
+    if dmr:
+        dmr_results = url_tester_ext(url)
+        if dap and dmr_results["dmr_test"].result == "pass":
+            dap_results = dap_tester(url)
+        else:
+            dap = False
+
+    test_results = {"dmr": dmr_results if dmr else "NA",
+                    "dap": dap_results if dap else "NA",
                     "netcdf4": url_tester_ext(url, '.dap.nc4') if nc4 else "NA"}
+
     return test_results
 
+
+def print_results(results):
+    print()
+    print("----- Results -----")
+    dmr_results = results["dmr"]
+    print(dmr_results["dmr_test"].result + " | " + str(dmr_results["dmr_test"].status) + " | dmr test")
+
+    if dmr_results["dmr_test"].result == "pass":
+        dap_results = results["dap"]
+        keys = dap_results.keys()
+        for k in keys:
+            print(dap_results[k].result + " | " + str(dap_results[k].status) + " | " + k)
+    print("/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/")
+
+
+def main():
```

Comment:
Is this needed anymore?

Reply:
not sure either way, gonna leave it in for now
there are a few things that im gonna add in the future that might need testing, so better to leave it in then have to remake it later